components/collections/src/codepointtrie/cptrie.rs

```diff
@@ -390,6 +439,75 @@ impl<'trie, T: TrieValue> CodePointTrie<'trie, T> {
         fast_index_val
     }
 
+    /// Returns the value that is associated with `code_point` in this [`CodePointTrie`]
+    /// assuming that the fast index path should be used.
+    ///
+    /// # Safety
+    ///
+    /// `code_point` must be at most `FAST_TYPE_FAST_INDEXING_MAX` if
+    /// the trie type is fast or at most `SMALL_TYPE_FAST_INDEXING_MAX`
+    /// if the trie type is small.
+    ///
+    /// # Panics
+    ///
+    /// When debug assertions are enabled, the above precondition and
+    /// slice access bounds are checked.
+    #[inline(always)]
+    unsafe fn get32_by_fast_index(&self, code_point: u32) -> T {
+        debug_assert!(
+            code_point
+                <= match self.header.trie_type {
+                    TrieType::Fast => FAST_TYPE_FAST_INDEXING_MAX,
+                    TrieType::Small => SMALL_TYPE_FAST_INDEXING_MAX,
+                }
+        );
+        let bit_prefix = (code_point as usize) >> FAST_TYPE_SHIFT;
+        debug_assert!(bit_prefix < self.index.len());
+        // SAFETY: The length of `self.index` has been checked
+        // in the constructor.
+        let base_offset_to_data: usize = usize::from(u16::from_unaligned(*unsafe {
+            self.index.as_ule_slice().get_unchecked(bit_prefix)
+        }));
```

Comment:
Suggestion: Style-wise, I wonder if we could write this as

```suggestion
        let base_offset_to_data: usize = usize::from(
            self.index.get(bit_prefix).unwrap_or_else(|| {
                debug_assert!(false, "!!! SAFETY INVARIANT VIOLATED");
                // SAFETY: The length of `self.index` has been checked
                // in the constructor.
                unsafe { core::hint::unreachable_unchecked() }
            })
        });
```

Then we have even more safety locality, and we don't need to use `as_ule_slice`.

Reply:
Both are UB, the former typically compiles to a `mov` (for the compiler to decide it is unreachable and elide it, it needs to know what the validity range is at compile time; which modern compilers only know in some situations, not ones like this)

The latter compiles to an unreachable hint which usually causes code deletion.

Reply:
Staying away from `core::hint::unreachable_unchecked()`.